File: webapps/api/backend/views/tasks.py
# Copyright (C) 2016 iNuron NV
#
# This file is part of Open vStorage Open Source Edition (OSE),
# as available from
#
#      http://www.openvstorage.org and
#      http://www.openvstorage.com.
#
# This file is free software; you can redistribute it and/or modify it
# under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
# as published by the Free Software Foundation, in version 3 as it comes
# in the LICENSE.txt file of the Open vStorage OSE distribution.
#
# Open vStorage is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY of any kind.

"""
Module for working with celery tasks
"""

from backend.decorators import load, log, required_roles, return_simple
from celery.task.control import inspect
from ovs.celery_run import celery
from rest_framework import status
from rest_framework import viewsets
from rest_framework.decorators import link
from rest_framework.permissions import IsAuthenticated
from rest_framework.response import Response


class TaskViewSet(viewsets.ViewSet):
    """
    Information about celery tasks
    """
    permission_classes = (IsAuthenticated,)
    prefix = r'tasks'
    base_name = 'tasks'

    @log()
    @required_roles(['read'])
    @return_simple()
    @load()
    def list(self):
        """
        Overview of active, scheduled, reserved and revoked tasks
        :return: Dict of all tasks
        :rtype: dict
        """
        inspector = inspect()
        return {'active': inspector.active(),
                'scheduled': inspector.scheduled(),
                'reserved': inspector.reserved(),
                'revoked': inspector.revoked()}

    @log()
    @required_roles(['read'])
    @return_simple()
    @load()
    def retrieve(self, pk):
        """
        Load information about a given task
        :param pk: Task identifier
        :type pk: str
        :return: Task metadata (status, result, ...)
        :rtype: dict
        """
        result = celery.AsyncResult(pk)
        result_status = result.status
        result_successful = result.successful()
        if result_successful is True:
            result_data = result.result
        else:
            result_data = str(result.result) if result.result is not None else None
        return {'id': result.id,
                'status': result_status,
                'successful': result_successful,
                'failed': result.failed(),
                'ready': result.ready(),
                'result': result_data}

    @link()
    @log()
    @required_roles(['read'])
    @return_simple()
    @load()
    def get(self, pk):
        """
        Gets a given task's result
        :param pk: Task identifier
        :type pk: str
        :return: Task result
        :rtype: dict
        """
        result = celery.AsyncResult(pk)
        return result.get()

File: webapps/frontend/lib/ovs/services/tasks.js
// Copyright (C) 2016 iNuron NV
//
// This file is part of Open vStorage Open Source Edition (OSE),
// as available from
//
//      http://www.openvstorage.org and
//      http://www.openvstorage.com.
//
// This file is free software; you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License v3 (GNU AGPLv3)
// as published by the Free Software Foundation, in version 3 as it comes
// in the LICENSE.txt file of the Open vStorage OSE distribution.
//
// Open vStorage is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY of any kind.
/*global define */
define([
    'jquery',
    'ovs/generic', 'ovs/api',
    'ovs/services/messaging'
], function($,
            generic, api,
            messaging) {
    "use strict";
    function Tasks() {
        var self = this;
        self.hooks       = {};
        self.taskIDQueue = [];

        // On task complete events, the task result should be loaded
        // @todo make this obsolete. The wait function should always resolve and not rely on messaging
        messaging.subscribe('TASK_COMPLETE', function(taskID) {
            if (self.hooks.hasOwnProperty(taskID)) {
                self.load(taskID)
                    .done(self.hooks[taskID].resolve)
                    .fail(self.hooks[taskID].reject);
            } else {
                var now = generic.getTimestamp(), i, newQueue = [];
                self.taskIDQueue.push({ id: taskID, timestamp: now });
                for (i = 0; i < self.taskIDQueue.length; i += 1) {
                    if (self.taskIDQueue[i].timestamp >= now - 10000) {
                        newQueue.push(self.taskIDQueue[i]);
                    }
                }
                self.taskIDQueue = newQueue;
            }
        });

        // Bind all prototype method to this instance. Used for backwards compatibility
        $.each(Tasks.prototype, function(key, value) {
            if (generic.isFunction(value)) {
                self[key] = value.bind(self)
            }
        })
    }
    Tasks.prototype = {
        /**
         * Wait until a task is resolved
         * Currently relies on messaging to resolve the task (see messaging.subscribe) in the constructor
         * @param taskID: ID of the task to wait for
         * @return {Promise<T>}
         */
        wait: function(taskID) {
            for (var i = 0; i < this.taskIDQueue.length; i += 1){
                var task = this.taskIDQueue[i];
                if (task.id === taskID) {
                    return this.load(taskID);
                }
            }
            this.hooks[taskID] = $.Deferred();
            return this.hooks[taskID].promise();
        },
        /**
         * Alternative to wait. Does not rely on messaging but instead longpolls on the task state
         * Might not be the better alternative as server side does not implement any longpolling like the messaging does
         * @param taskID: ID of the task to wait for
         * @return {Promise<T>}
         */
        waitPoll: function(taskID) {
            var self = this;
            return api.get('tasks/' + taskID)
                .then(function(data) {
                    if (!data.ready) {
                        return generic.delay(1000).then(function() {
                            return self.waitPoll(taskID)
                        })
                    }
                    if (data.successful) {
                        return data.result;
                    } else {
                        throw data.result
                    }
                })
        },
        /**
         * Load task results
         * @param taskID: ID of the task
         * @return {Promise<T>}
         */
        load: function(taskID) {
            return api.get('tasks/' + taskID)
                .then(function(data) {
                    if (data.successful) {
                        return data.result;
                    } else {
                        throw data.result
                    }})
        },
        /**
         * Validate all tasks
         * Checks the current hooks and resolves where possible
         */
        validateTasks: function() {
            $.each(this.hooks, function(taskID, deferred) {
                api.get('tasks/' + taskID)
                    .done(function(data) {
                        if (data.ready === true) {
                            if (data.successful === true) {
                                deferred.resolve(data.result);
                            } else {
                                deferred.reject(data.result);
                            }
                        }
                    });
            });
        }
    };

    return new Tasks()
});
